containerd/containerd.go

```diff
@@ -335,3 +337,50 @@ func (d *Driver) getTask(container containerd.Container) (containerd.Task, error
 
 	return container.Task(ctxWithTimeout, cio.Load)
 }
+
+// WithMemoryLimits sets the `Linux.LinuxResources.Memory.Limit` section to the
+// `limit` specified if the `Linux` section is not `nil`. Additionally sets the
+// `Windows.WindowsResources.Memory.Limit` section if the `Windows` section is
+// not `nil`.
+//
+// If hard limit is set on Windows, the soft limit is ignored.
+func WithMemoryLimits(soft, hard uint64) oci.SpecOpts {
+	return func(_ context.Context, _ oci.Client, _ *containers.Container, s *oci.Spec) error {
+		if s.Linux != nil {
+			if s.Linux.Resources == nil {
+				s.Linux.Resources = &specs.LinuxResources{}
+			}
+			if s.Linux.Resources.Memory == nil {
+				s.Linux.Resources.Memory = &specs.LinuxMemory{}
+			}
+
+			if hard > 0 {
+				hl := int64(hard)
+				s.Linux.Resources.Memory.Limit = &hl
+
+				sl := int64(soft)
+				s.Linux.Resources.Memory.Reservation = &sl
+
+			} else {
+				l := int64(soft)
+				s.Linux.Resources.Memory.Limit = &l
+			}
+		}
+
+		if s.Windows != nil {
```

Comment:
Do we need `Windows` checks here? Currently `containerd-driver` is only written with `Linux` support. There is no `Windows` logic in the codebase. We may think of how to extend this to `Windows` in future, but that would be a much broader discussion.

Another factor is `containerd` itself is mostly designed for Linux as a first class citizen in mind. `Windows+containers` still need a lot of work, and AFAIK it's not there (production-ready) yet.

I am thinking, just keep the existing `containerd-driver` codebase for `Linux` only. and we can discuss about windows in future.